johnzon-jsonb/src/main/java/org/apache/johnzon/jsonb/JsonbPolymorphismHandler.java

```diff
@@ -0,0 +1,155 @@
+            validateOnlyOneParentWithTypeInfo(current);
+
+            JsonbTypeInfo typeInfo = Meta.getDirectAnnotation(current, JsonbTypeInfo.class);
+            if (typeInfo != null) {
+                if (otherProperties.contains(typeInfo.key())) {
+                    throw new JsonbException("JsonbTypeInfo key '" + typeInfo.key() + "' collides with other properties in json");
+                }
+
+                if (entries.stream().anyMatch(entry -> Objects.equals(entry.getKey(), typeInfo.key()))) {
+                    throw new JsonbException("JsonbTypeInfo key '" + typeInfo.key() + "' found more than once in type hierarchy of " + clazz.getName());
+                }
+
+                String bestMatchingAlias = null;
+                for (JsonbSubtype subtype : typeInfo.value()) {
+                    if (subtype.type().isAssignableFrom(clazz)) {
+                        bestMatchingAlias = subtype.alias();
+
+                        if (clazz == subtype.type()) { // Exact match found, no need to continue further
+                            break;
+                        }
+                    }
+                }
+
+                if (bestMatchingAlias != null) {
+                    entries.add(0, Map.entry(typeInfo.key(), bestMatchingAlias));
+                }
+            }
+
+            List<Class<?>> parentClassesWithTypeInfo = getParentClassesWithTypeInfo(current);
+            current = parentClassesWithTypeInfo.isEmpty() ? null : parentClassesWithTypeInfo.get(0);
+        }
+
+        return entries;
+    }
+
+    public Class<?> getTypeToDeserialize(JsonObject jsonObject, Class<?> clazz) {
+        validateJsonbTypeInfo(clazz);
+        validateOnlyOneParentWithTypeInfo(clazz);
+
+        JsonbTypeInfo typeInfo = Meta.getDirectAnnotation(clazz, JsonbTypeInfo.class);
+        if (typeInfo == null || !jsonObject.containsKey(typeInfo.key())) {
+            return clazz;
+        }
+
+        JsonValue typeValue = jsonObject.get(typeInfo.key());
+        if (!(typeValue instanceof JsonString)) {
+            throw new JsonbException("Property '" + typeInfo.key() + "' isn't a String, resolving "
+                    + "JsonbSubtype is impossible");
+        }
+
+        String typeValueString = ((JsonString) typeValue).getString();
+        for (JsonbSubtype subtype : typeInfo.value()) {
+            if (subtype.alias().equals(typeValueString)) {
+                return subtype.type();
+            }
+        }
+
+        throw new JsonbException("No JsonbSubtype found for alias '" + typeValueString + "' on " + clazz.getName());
+    }
+
+    /**
+     * Validates that only one parent class (superclass + interfaces) has {@link JsonbTypeInfo} annotation
+     *
+     * @param classToValidate class to validate
+     * @throws JsonbException validation failed
+     */
+    private void validateOnlyOneParentWithTypeInfo(Class<?> classToValidate) {
+        if (getParentClassesWithTypeInfo(classToValidate).size() > 1) {
```

Comment:
this looks called too often (at least twice per type if I'm not mistaken, maybe reuse the same instance for both needs?)